flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopConfigLoader.java

```diff
@@ -0,0 +1,134 @@
+	private final String flinkShadingPrefix;
+
+	/** Flink's configuration object. */
+	private Configuration flinkConfig;
+
+	/** Hadoop's configuration for the file systems, lazily initialized. */
+	private org.apache.hadoop.conf.Configuration hadoopConfig;
+
+	public HadoopConfigLoader(
+		@Nonnull String[] flinkConfigPrefixes,
+		@Nonnull String[][] mirroredConfigKeys,
+		@Nonnull String hadoopConfigPrefix,
+		@Nonnull Set<String> configKeysToShade,
+		@Nonnull String flinkShadingPrefix) {
+		this.flinkConfigPrefixes = flinkConfigPrefixes;
+		this.mirroredConfigKeys = mirroredConfigKeys;
+		this.hadoopConfigPrefix = hadoopConfigPrefix;
+		this.configKeysToShade = configKeysToShade;
+		this.flinkShadingPrefix = flinkShadingPrefix;
+	}
+
+	public void setFlinkConfig(Configuration config) {
+		flinkConfig = config;
+		hadoopConfig = null;
+	}
+
+	/** get the loaded Hadoop config (or fall back to one loaded from the classpath). */
+	public org.apache.hadoop.conf.Configuration getOrLoadHadoopConfig() {
+		org.apache.hadoop.conf.Configuration hadoopConfig = this.hadoopConfig;
+		if (hadoopConfig == null) {
+			if (flinkConfig != null) {
+				hadoopConfig = mirrorCertianHadoopConfig(loadHadoopConfigFromFlink());
+			}
+			else {
+				LOG.warn("The factory has not been configured prior to loading the S3 file system."
+					+ " Using Hadoop configuration from the classpath.");
+				hadoopConfig = new org.apache.hadoop.conf.Configuration();
+			}
+		}
+		this.hadoopConfig = hadoopConfig;
+		return hadoopConfig;
+	}
+
+	// add additional config entries from the Flink config to the Hadoop config
+	private org.apache.hadoop.conf.Configuration loadHadoopConfigFromFlink() {
+		org.apache.hadoop.conf.Configuration hadoopConfig = HadoopUtils.getHadoopConfiguration(flinkConfig);
+		for (String key : flinkConfig.keySet()) {
+			for (String prefix : flinkConfigPrefixes) {
+				if (key.startsWith(prefix)) {
+					String value = flinkConfig.getString(key, null);
+					String newKey = hadoopConfigPrefix + key.substring(prefix.length());
+					String newValue = fixHadoopConfig(key, flinkConfig.getString(key, null));
+					hadoopConfig.set(newKey, newValue);
+
+					LOG.debug("Adding Flink config entry for {} as {}={} to Hadoop config", key, newKey, value);
+				}
+			}
+		}
+		return hadoopConfig;
+	}
+
+	// mirror certain keys to make use more uniform across implementations
+	// with different keys
+	private org.apache.hadoop.conf.Configuration mirrorCertianHadoopConfig(
+		org.apache.hadoop.conf.Configuration hadoopConfig) {
+		for (String[] mirrored : mirroredConfigKeys) {
+			String value = hadoopConfig.get(mirrored[0], null);
+			if (value != null) {
+				hadoopConfig.set(mirrored[1], value);
+			}
+		}
+		return hadoopConfig;
+	}
+
+	private String fixHadoopConfig(String key, String value) {
+		return key != null && configKeysToShade.contains(key) ?
+			shadeAwsClassConfig(value) : value;
+	}
+
+	private String shadeAwsClassConfig(String awsCredProvider) {
```

Comment:
This is not tied to AWS or the credentials provider, I think.

Reply:
true, overlooked